### PR TITLE
Expose `get_module_export` from `Caller`

### PR DESCRIPTION
**Pull request by richardpringle**
Not necessarily trying to get this merged, just using it as an example for my comment [here](https://github.com/bytecodealliance/wasmtime/issues/6848#issuecomment-2447900164)

### CODE REVIEW

**Comment by alexcrichton** [comment]
What I might recommend for your use case is to store a `TypedFunc<P, R>` within the `T` of your `Store<T>`. That'd probably start as `Option<TypedFunc<...>>` and be `None` on store construction. After instantiation you'd load the function, set it in the store, and then all your host functions have access to that allocation function.

Alternatively if you want the absolute fastest route what I think you'd do is:

* Load a `Module`
* Load the [export and type-check it manually](https://docs.rs/wasmtime/latest/wasmtime/struct.Module.html#method.get_export)
* Load the [`ModuleExport`](https://docs.rs/wasmtime/latest/wasmtime/struct.Module.html#method.get_export_index)
* Use `InstancePre` to instantiate quickly
* Use the [`ModuleExport` to get the exported function](https://docs.rs/wasmtime/latest/wasmtime/struct.Instance.html#method.get_module_export)
* Use [`TypedFunc::new_unchecked`](https://docs.rs/wasmtime/latest/wasmtime/struct.TypedFunc.html#method.new_unchecked) with the resulting function

That front-loads as much as possible at the cost of safety, so I'd recommend using `.get_typed_func` if possible after the initial instantiation of the module.

File: crates/wasmtime/src/runtime/func.rs [hunk 1]
```diff
@@ -2109,6 +2109,26 @@ impl<T> Caller<'_, T> {
...
+    /// string lookups where possible. [`ModuleExport`]s can be obtained by calling
+    /// [`Module::get_export_index`] on the [`Module`] that an instance was instantiated with.
+    ///
+    /// This method will search the module for an export with a matching entity index and return
+    /// the value, if found.
+    ///
+    /// Returns `None` if there was no export with a matching entity index.
+    /// # Panics
+    ///
+    /// Panics if `store` does not own this instance.
```
Comment (richardpringle) [inline crates/wasmtime/src/runtime/func.rs:2124]: Docs are effectively a copy/paste of `Instance::get_module_export`. Happy to take recommendations or if you want to make a suggestion, I will definitely accept it. I think `see [Instance::get_module_export] for details` would probably also be fine so long as it's clear that the `caller` has access to the instance (wasn't clear to me without looking at the code).

**Comment by richardpringle** [comment]
> Looks good to me! Mind adding a small smoke test too?

@alexcrichton not sure if this is what you meant by a smoke test, but it works as one, haha 😅 .

**Review by alexcrichton — APPROVED** [review]
Thanks!

**Pull request merged**